File: dotnet/examples/Connection/Resilience/ReconnectionStrategy.cs
/**
 * Copyright © 2023 - 2024 Diffusion Data Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
using System;
using System.Threading;
using System.Threading.Tasks;
using PushTechnology.ClientInterface.Client.Factories;
using PushTechnology.ClientInterface.Client.Session;
using PushTechnology.ClientInterface.Client.Session.Reconnection;
using static System.Console;
using static PushTechnology.ClientInterface.Examples.Program;

namespace PushTechnology.ClientInterface.Examples.Connection.Resilience
{
    public sealed class ReconnectionStrategy : Example
    {
        public override async Task Run(CancellationToken cancellationToken, string[] args)
        {
            string serverUrl = args[0];

            // Set the maximum amount of time the client will try and reconnect for to 10 minutes.
            int maximumTimeoutDuration = 1000 * 60 * 10;
            var reconnectionStrategy = new SessionReconnectionStrategy();

            var session = Diffusion.Sessions
                .Principal("admin")
                .Credentials(Diffusion.Credentials.Password("password"))
                .ReconnectionTimeout(maximumTimeoutDuration)
                .ReconnectionStrategy(reconnectionStrategy)
                .Open(serverUrl);

            WriteLine($"Connected. Session Identifier: {session.SessionId}.");

            // Insert work here...

            session.Close();
        }

        /// <summary>
        /// This class is used to reconnect to the server when the session is disconnected unexpectedly.
        /// </summary>
        private class SessionReconnectionStrategy : IReconnectionStrategy
        {
            public int Retries { get; set; }

            public SessionReconnectionStrategy() => Retries = 0;

            public async Task PerformReconnection(IReconnectionAttempt reconnection)
            {
                if (Retries > 10) 
                {
                    // Abort after 10 attempts.
                    throw new Exception();
                }
                else
                {
                    Retries++;

                    // Retry after 3 seconds.
                    await Task.Delay(TimeSpan.FromMilliseconds(3000));

                    WriteLine("Attempting to reconnect...");
                   
                    reconnection.Start();
                }
            }
        }
    }
}

File: c/examples/connection/resilience/reconnection_strategy.c
/**
 * Copyright © 2024 DiffusionData Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * This example is written in C99. Please use an appropriate C99 capable compiler
 */

#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#include "diffusion.h"
#include "utils.h"
MUTEX_DEF

void run_example(
    const char *url,
    const char *principal,
    CREDENTIALS_T *credentials)
{
    RECONNECTION_STRATEGY_T reconnection_strategy = {
        .retry_count = 3,
        .retry_delay = 1000
    };

    DIFFUSION_ERROR_T error = { 0 };
    SESSION_T *session = session_create(
        url, principal, credentials, NULL, &reconnection_strategy, &error
    );

    // Insert work here

    session_close(session, NULL);
    session_free(session);
}